service/rpc/share.go

```diff
@@ -76,3 +62,104 @@ func (h *Handler) handleSharesByNamespaceRequest(w http.ResponseWriter, r *http.
 		log.Errorw("serving request", "endpoint", namespacedSharesEndpoint, "err", err)
 	}
 }
+
+func (h *Handler) handleDataByNamespaceRequest(w http.ResponseWriter, r *http.Request) {
+	height, nID, err := parseGetByNamespaceArgs(r)
+	if err != nil {
+		writeError(w, http.StatusBadRequest, namespacedDataEndpoint, err)
+		return
+	}
+	shares, headerHeight, err := h.getShares(r.Context(), height, nID)
+	if err != nil {
+		writeError(w, http.StatusInternalServerError, namespacedDataEndpoint, err)
+		return
+	}
+	data, err := dataFromShares(shares)
+	if err != nil {
+		writeError(w, http.StatusInternalServerError, namespacedDataEndpoint, err)
+		return
+	}
+	resp, err := json.Marshal(&NamespacedDataResponse{
+		Data:   data,
+		Height: uint64(headerHeight),
+	})
+	if err != nil {
+		writeError(w, http.StatusInternalServerError, namespacedDataEndpoint, err)
+		return
+	}
+	_, err = w.Write(resp)
+	if err != nil {
+		log.Errorw("serving request", "endpoint", namespacedDataEndpoint, "err", err)
+	}
+}
+
+func (h *Handler) getShares(ctx context.Context, height uint64, nID namespace.ID) ([]share.Share, int64, error) {
+	// get header
+	var (
+		err    error
+		header *header.ExtendedHeader
+	)
+	switch height {
+	case 0:
+		header, err = h.header.Head(ctx)
+	default:
+		header, err = h.header.GetByHeight(ctx, height)
+	}
+	if err != nil {
+		return nil, 0, err
+	}
+	// perform request
+	shares, err := h.share.GetSharesByNamespace(ctx, header.DAH, nID)
+	return shares, header.Height, err
+}
+
+func dataFromShares(shares []share.Share) ([][]byte, error) {
+	for i := range shares {
+		insertDelimiter(shares[i])
```

Comment:
@evan-forbes I was testing this mostly on blocks from `devnet-2`, generated with older node/core/app.
Do I understand correctly, that this step will no longer be needed with up-to-date PFD processing?

Reply:
I don't think you should have had to add that at all, even for old devnet!

it should just work like this https://github.com/celestiaorg/dalc/blob/6589e5dc61037646c663f86bd97e5036c56ec77e/server/server.go#L130-L138

https://github.com/celestiaorg/dalc/blob/6589e5dc61037646c663f86bd97e5036c56ec77e/server/server_test.go#L14-L55

Reply:
It was always hidden behind other issues (like random share ordering).